replica/mutation_dump.cc

```diff
@@ -0,0 +1,531 @@
+            tracing::trace_state_ptr ts, db::timeout_clock::time_point timeout)
+        : _db(db)
+        , _schema(std::move(schema))
+        , _ts(std::move(ts))
+        , _timeout(timeout)
+        , _cmd(
+                _schema->id(),
+                _schema->version(),
+                make_slice(),
+                query::max_result_size(query::result_memory_limiter::maximum_result_size),
+                query::tombstone_limit(1000),
+                query::row_limit::max,
+                query::partition_limit(1000),
+                gc_clock::now(),
+                tracing::make_trace_info(_ts),
+                query_id::create_random_id(),
+                query::is_first_page::yes)
+    {
+        _prs.reserve(prs.size());
+        std::copy(prs.begin(), prs.end(), std::back_inserter(_prs));
+    }
+    future<std::optional<dht::decorated_key>> operator()() {
+        while (_dks.empty() && !_prs.empty()) {
+            co_await fill_dk_buffer_from_current_range();
+        }
+        if (_dks.empty()) {
+            co_return std::nullopt;
+        }
+        auto ret = std::optional(std::move(_dks.front()));
+        _dks.pop_front();
+        co_return std::move(ret);
+    }
+};
+
+noncopyable_function<future<std::optional<dht::decorated_key>>()>
+make_partition_key_generator(distributed<replica::database>& db, schema_ptr schema, const dht::partition_range_vector& prs,
+        tracing::trace_state_ptr ts, db::timeout_clock::time_point timeout) {
+    if (prs.size() == 1 && prs.front().is_singular()) {
+        auto dk_opt = std::optional(prs.front().start()->value().as_decorated_key());
+        return [dk_opt = std::move(dk_opt)] () mutable {
+            return make_ready_future<std::optional<dht::decorated_key>>(std::exchange(dk_opt, std::nullopt));
+        };
+    }
+    return multi_range_partition_generator(db, std::move(schema), prs, std::move(ts), timeout);
+}
+
+} // anonymous namespace
+
+schema_ptr generate_output_schema_from_underlying_schema(schema_ptr underlying_schema) {
+    auto sb = schema_builder(underlying_schema->ks_name(), format("{}_$mutation_fragments", underlying_schema->cf_name()));
+    // partition key
+    for (const auto& pk_col : underlying_schema->partition_key_columns()) {
+        sb.with_column(pk_col.name(), pk_col.type, column_kind::partition_key);
+    }
+    // clustering key
+    sb.with_column("mutation_source", utf8_type, column_kind::clustering_key);
+    sb.with_column("partition_region", byte_type, column_kind::clustering_key);
+    for (const auto& ck_col : underlying_schema->clustering_key_columns()) {
+        sb.with_column(ck_col.name(), ck_col.type, column_kind::clustering_key);
+    }
+    sb.with_column("position_weight", byte_type, column_kind::clustering_key);
+    // regular columns
+    sb.with_column("mutation_fragment_kind", utf8_type);
+    sb.with_column("metadata", utf8_type);
+    sb.with_column("value", utf8_type);
+    return sb.build();
```

Comment:
I think we should set the uuid as a transformation of the input table_schema_version (say xor the lower bits with some pattern). This way, g_o_s_f_u_s() always returns the same schema if given the same input. May need to stop randomization of the returned schema's table_schema_version. Should test it too - calling gosfus twice on the same input should result in identical, not just equivalent schemas.

Reply:
Or maybe better: transform table_id into table_id, and version into version.

Reply:
Done in v4.